classes/question_ui_metadata_extractor.php

```diff
@@ -86,22 +86,60 @@ public function extract(): question_metadata {
             }
         }
 
+        $namemap = [];
         /** @var DOMElement $element */
         foreach (
-            $this->xpath->query(
-                '//*[self::xhtml:input or self::xhtml:select or self::xhtml:textarea or self::xhtml:button]'
-            ) as $element
+            $this->xpath->query('(//xhtml:button | //xhtml:input | //xhtml:select | //xhtml:textarea)[@name]') as $element
         ) {
             $name = $element->getAttribute('name');
-            if ($name) {
-                $this->metadata->expecteddata[$name] = PARAM_RAW;
+            $this->metadata->expecteddata[$name] = PARAM_RAW;
 
-                if ($element->hasAttribute('required')) {
-                    $this->metadata->requiredfields[] = $name;
-                }
+            if ($element->hasAttribute('required')) {
+                $this->metadata->requiredfields[] = $name;
+            }
+
+            // Collect all field names whose values are a json encoded list.
+            if (in_array($name, $this->metadata->jsonfieldnames)) {
+                continue;
+            }
+
+            $type = $element->getAttribute('type') ?: 'text';
+            $value = $element->hasAttribute('value') ? $element->getAttribute('value') : 'on';
+
+            if (!isset($namemap[$name])) {
+                // This name has not been used yet by other elements.
+                $namemap[$name] = [$element, [$value]];
+                continue;
+            }
+
+            // Get the type and values of the other element(s) with the same name.
+            [$other, &$values] = $namemap[$name];
+            $othertype = $other->getAttribute('type') ?: 'text';
+
+            if (!in_array($othertype, ['checkbox', 'radio'])) {
+                // Duplicate names are not allowed for other elements.
+                $this->metadata->jsonfieldnames[] = $name;
+                continue;
+            }
+
+            // Check that the types match and the values is unique.
+            if ($othertype !== $type || in_array($value, $values)) {
+                $this->metadata->jsonfieldnames[] = $name;
+            } else {
+                $values[] = $value;
             }
         }
 
+        $this->metadata->jsonfieldnames = array_merge(
+            $this->metadata->jsonfieldnames,
+            array_keys(array_filter($namemap, function ($elementvalues, $name) {
+                [$element, $values] = $elementvalues;
+                $aremultiplecheckboxes = $element->getAttribute('type') === 'checkbox' && count($values) > 1;
+                $ismultiselect = $element->tagName === 'select' && $element->hasAttribute('multiple');
+                return !in_array($name, $this->metadata->jsonfieldnames) && ($aremultiplecheckboxes || $ismultiselect);
+            }, ARRAY_FILTER_USE_BOTH))
+        );
```

Comment:
Ich wäre ganz stark dafür, das zu entfunktionalisieren: 
```suggestion
        foreach ($namemap as $name => [$element, $values]) {
            $aremultiplecheckboxes = $element->getAttribute('type') === 'checkbox' && count($values) > 1;
            $ismultiselect = $element->tagName === 'select' && $element->hasAttribute('multiple');

            if (!in_array($name, $this->metadata->jsonfieldnames) && ($aremultiplecheckboxes || $ismultiselect)) {
                $this->metadata->jsonfieldnames[] = $name;
            }
        }
```

Oder so :smile: